File: 278-first-bad-version/first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
         int startIndex = 1; // camelCase (Kyunki ginti 1 se n tak hai)
        int endIndex = n;   // camelCase (Isliye n-1 nahi kiya)
        
        while (startIndex <= endIndex) {
            // Integer overflow se bachne ke liye safe formula
            int midIndex = startIndex + (endIndex - startIndex) / 2; // camelCase
            
            // Agar yeh version kharab hai, toh ho sakta hai pehla kharab aur piche ho
            if (isBadVersion(midIndex)) {
                endIndex = midIndex - 1; // Left side check karne jao
            } 
            // Agar yeh version sahi hai, toh pakka pehla kharab aage hi hoga
            else {
                startIndex = midIndex + 1; // Right side check karne jao
            }
        }
        
        // Loop break hone par startIndex automatically pehle bad version par rukta hai
        return startIndex;
    }
}

File: 278-first-bad-version/first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
         int start_index = 1; // snake_case (Kyunki ginti 1 se n tak hai)
        int end_index = n;   // snake_case (Isliye n-1 nahi kiya)
        
        while (start_index <= end_index) {
            // Integer overflow se bachne ke liye safe formula
            int mid_index = start_index + (end_index - start_index) / 2; // snake_case
            
            // Agar bad version mil gaya, toh aur piche check karo left me
            if (isBadVersion(mid_index)) {
                end_index = mid_index - 1;
            } 
            // Agar sahi version hai, toh aage right me jao
            else {
                start_index = mid_index + 1;
            }
        }
        
        // Loop break hone par start_index hi sabse pehla bad version hota hai
        return start_index;
    }
};